lib/src/Uncertainty/Algorithm/WeightedExperiments/TensorProductExperiment.cxx

```diff
@@ -0,0 +1,165 @@
+//                         -*- C++ -*-
+/**
+ *  @brief Abstract top-level view of a TensorProductExperiment
+ *
+ *  Copyright 2005-2021 Airbus-EDF-IMACS-ONERA-Phimeca
+ *
+ *  This library is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/TensorProductExperiment.hxx"
+#include "openturns/Tuples.hxx"
+#include "openturns/Exception.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/Log.hxx"
+#include <iostream>
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(TensorProductExperiment)
+
+typedef Collection< Point > PointCollection;
+
+/* Default constructor */
+TensorProductExperiment::TensorProductExperiment()
+  : WeightedExperimentImplementation()
+  , collection_(0)
+  , nodes_(0, 0)
+  , isAlreadyComputedNodesAndWeights_(false)
+{
+  // Nothing to do
+}
+
+/* Constructor with parameters */
+TensorProductExperiment::TensorProductExperiment(const WeightedExperimentCollection & collection)
+  : WeightedExperimentImplementation()
+  , collection_(collection)
+  , nodes_(0, 0)
+  , isAlreadyComputedNodesAndWeights_(false)
+{
+  // Nothing to do
+}
+
+/* Virtual constructor */
+TensorProductExperiment * TensorProductExperiment::clone() const
+{
+  return new TensorProductExperiment(*this);
+}
+
+/* String converter */
+String TensorProductExperiment::__repr__() const
+{
+  OSS oss;
+  oss << "class=" << GetClassName()
+    << " name=" << getName()
+    << " collection=" << collection_;
+  return oss;
+}
+
+Bool TensorProductExperiment::hasUniformWeights() const
+{
+  return false;
```

Comment:
it depends on the various marginals, isn't it?

Reply:
Oups: copy-paste does not always work... You are correct. For example, if we use a MonteCarloExperiment, the returned value should be True. I have to fix this. To implement it, I must call the `hasUniformWeights` method of each marginal experiment: the returned value is `true` if all of them are.

Reply:
Ok. Fixed.